SU2_CFD/src/fluid/CDataDrivenFluid.cpp

```diff
@@ -137,44 +178,61 @@ void CDataDrivenFluid::MapInputs_to_Outputs() {
-  Density = rho;
-  StaticEnergy = e;
 
-  /*--- Clip density and energy values to prevent extrapolation. ---*/
-  Density = min(rho_max, max(rho_min, Density));
-  StaticEnergy = min(e_max, max(e_min, StaticEnergy));
-
-  Evaluate_Dataset(Density, StaticEnergy);
+  AD::StartPreacc();
+  AD::SetPreaccIn(rho);
+  AD::SetPreaccIn(e);
 
-  /*--- Compute speed of sound. ---*/
-  auto blue_term = (dsdrho_e * (2 - rho * pow(dsde_rho, -1) * d2sdedrho) + rho * d2sdrho2);
-  auto green_term = (-pow(dsde_rho, -1) * d2sde2 * dsdrho_e + d2sdedrho);
+  /*--- Compute thermodynamic state based on density and energy. ---*/
+  
+  Density = max(min(rho, rho_max), rho_min);
+  StaticEnergy = max(min(e, e_max), e_min);
 
-  SoundSpeed2 = -rho * pow(dsde_rho, -1) * (blue_term - rho * green_term * (dsdrho_e / dsde_rho));
+  Evaluate_Dataset(Density, StaticEnergy);
 
+  su2double rho_2 = Density * Density;
   /*--- Compute primary flow variables. ---*/
-  Temperature = 1.0 / dsde_rho;
-  Pressure = -pow(rho, 2) * Temperature * dsdrho_e;
-  Density = rho;
-  StaticEnergy = e;
-  Enthalpy = e + Pressure / rho;
+  Temperature = pow(dsde_rho, -1);
+  Pressure = -rho_2 * Temperature * dsdrho_e;
+  Enthalpy = StaticEnergy + Pressure / Density;
 
   /*--- Compute secondary flow variables ---*/
-  dTde_rho = -pow(dsde_rho, -2) * d2sde2;
-  dTdrho_e = -pow(dsde_rho, -2) * d2sdedrho;
+  dTde_rho = -Temperature * Temperature * d2sde2;
+  dTdrho_e = -Temperature * Temperature * d2sdedrho;
 
-  dPde_rho = -pow(rho, 2) * (dTde_rho * dsdrho_e + Temperature * d2sdedrho);
-  dPdrho_e = -2 * rho * Temperature * dsdrho_e - pow(rho, 2) * (dTdrho_e * dsdrho_e + Temperature * d2sdrho2);
+  /*--- Compute speed of sound. ---*/
+  su2double blue_term = (dsdrho_e * (2 - Density * Temperature * d2sdedrho) + Density * d2sdrho2);
+  su2double green_term = (-Temperature * d2sde2 * dsdrho_e + d2sdedrho);
+
+  SoundSpeed2 = -Density * Temperature * (blue_term - Density * green_term * (dsdrho_e / dsde_rho));
+
+  dPde_rho = -rho_2 * Temperature * (-Temperature * (d2sde2 * dsdrho_e) + d2sdedrho);
+  dPdrho_e = - Density * Temperature * (dsdrho_e * (2 - Density * Temperature * d2sdedrho) + Density * d2sdrho2);
 
   /*--- Compute enthalpy and entropy derivatives required for Giles boundary conditions. ---*/
-  dhdrho_e = -Pressure * pow(rho, -2) + dPdrho_e / rho;
-  dhde_rho = 1 + dPde_rho / rho;
+  dhdrho_e = -Pressure * (1 / rho_2) + dPdrho_e / Density;
+  dhde_rho = 1 + dPde_rho / Density;
 
+  /*--- Compute specific heat at constant volume and specific heat at constant pressure. ---*/
+  Cv = 1 / dTde_rho;
   dhdrho_P = dhdrho_e - dhde_rho * (1 / dPde_rho) * dPdrho_e;
   dhdP_rho = dhde_rho * (1 / dPde_rho);
   dsdrho_P = dsdrho_e - dPdrho_e * (1 / dPde_rho) * dsde_rho;
   dsdP_rho = dsde_rho / dPde_rho;
+
+  su2double drhode_p = -dPde_rho/dPdrho_e;
+  su2double dTde_p = dTde_rho + dTdrho_e*drhode_p;
+  su2double dhde_p = dhde_rho + drhode_p*dhdrho_e;
+  Cp = dhde_p / dTde_p;
```

Comment:
Are Cv and Cp local variables? Why are they not preaccumulated? Do you have any adjoint verification cases yet to be playing with preaccumulation?

Reply:
I initially copied the preaccumulation from the Peng-Robinson fluid model. I added some unit tests for fluid model consistency and I'm planning to add some AD unit tests as well.

Reply:
I ran a gradient validation study where I found the maximum difference between the AD gradients and FD gradients to be 0.14%. @joshkellyjak and I did a preaccumulation debugging session together where we identified several instances in the CDataDrivenFluid class where the preaccumulation tag was wrong. After fixing those, I ran the AD again, but the percentage difference in AD gradients was in the order of 1e-5.